crates/forge/src/multi_runner.rs

```diff
@@ -281,72 +282,47 @@ impl MultiContractRunnerBuilder {
         // create a mapping of name => (abi, deployment code, Vec<library deployment code>)
         let mut deployable_contracts = DeployableContracts::default();
 
-        fn unique_deps(deps: Vec<ResolvedDependency>) -> Vec<ResolvedDependency> {
-            let mut filtered = Vec::new();
-            let mut seen = HashSet::new();
-            for dep in deps {
-                if !seen.insert(dep.id.clone()) {
-                    continue
-                }
-                filtered.push(dep);
+        let artifact_contracts = ArtifactContracts::from_iter(contracts.clone());
+
+        let linker = Linker::new(root.as_ref(), artifact_contracts);
+
+        for (id, contract) in contracts {
+            let abi = contract.abi.as_ref().ok_or_eyre("we should have an abi by now")?;
+
+            let LinkOutput { contracts, libs_to_deploy, .. } =
+                linker.link_with_nonce_or_address(Default::default(), evm_opts.sender, 1, &id)?;
+
+            let linked_contract = contracts.get(&id).unwrap().clone();
+
+            // get bytes if deployable, else add to known contracts and continue.
+            // interfaces and abstract contracts should be known to enable fuzzing of their ABI
+            // but they should not be deployable and their source code should be skipped by the
+            // debugger and linker.
+            let Some(bytecode) = linked_contract
+                .bytecode
+                .and_then(|b| b.object.into_bytes())
+                .filter(|b| !b.is_empty())
+            else {
+                known_contracts.insert(id.clone(), (abi.clone(), vec![]));
+                continue;
+            };
+
+            // if it's a test, add it to deployable contracts
+            if abi.constructor.as_ref().map(|c| c.inputs.is_empty()).unwrap_or(true) &&
+                abi.functions().any(|func| func.name.is_test() || func.name.is_invariant_test())
+            {
+                deployable_contracts.insert(id.clone(), (abi.clone(), bytecode, libs_to_deploy));
             }
 
-            filtered
+            linked_contract
+                .deployed_bytecode
+                .and_then(|d_bcode| d_bcode.bytecode)
+                .and_then(|bcode| bcode.object.into_bytes())
```

Comment:
I think this can be simplified with `.get_deployed_bytecode_bytes().cloned()`

we could also clean this up on foundry-compilers so that this function no longer returns Cow, not useful anyway